packages/next/src/server/request/params.ts

```diff
@@ -1 +1,472 @@
+  const prefix = route ? ` In route ${route} a ` : 'A '
+  console.error(
+    `${prefix}param property was accessed directly with ${expression}. \`params\` is now a Promise and should be awaited before accessing properties of the underlying params object. In this version of Next.js direct access to param properties is still supported to facilitate migration but in a future version you will be required to await \`params\`. If this use is inside an async function await it. If this use is inside a synchronous function then convert the function to async or await it from outside this function and pass the result in.`
+  )
+}
+
+function warnForEnumeration(
+  route: undefined | string,
+  missingProperties: Array<string>
+) {
+  const prefix = route ? ` In route ${route} ` : ''
+  if (missingProperties.length) {
+    const describedMissingProperties =
+      describeListOfPropertyNames(missingProperties)
+    console.error(
+      `${prefix}params are being enumerated incompletely with \`{...params}\`, \`Object.keys(params)\`, or similar. The following properties were not copied: ${describedMissingProperties}. \`params\` is now a Promise, however in the current version of Next.js direct access to the underlying params object is still supported to facilitate migration to the new type. param names that conflict with Promise properties cannot be accessed directly and must be accessed by first awaiting the \`params\` promise.`
+    )
+  } else {
+    console.error(
+      `${prefix}params are being enumerated with \`{...params}\`, \`Object.keys(params)\`, or similar. \`params\` is now a Promise, however in the current version of Next.js direct access to the underlying params object is still supported to facilitate migration to the new type. You should update your code to await \`params\` before accessing its properties.`
+    )
+  }
+}
+
+function describeListOfPropertyNames(properties: Array<string>) {
+  switch (properties.length) {
+    case 0:
+      throw new InvariantError(
+        'Expected describeListOfPropertyNames to be called with a non-empty list of strings.'
+      )
+    case 1:
+      return `\`${properties[0]}\``
+    case 2:
+      return `\`${properties[0]}\` and \`${properties[1]}\``
+    default: {
+      let description = ''
+      for (let i = 0; i < properties.length - 1; i++) {
+        description += `\`${properties[i]}\`, `
+      }
+      description += `, and \`${properties[properties.length - 1]}\``
+      return description
+    }
+  }
+}
+
+// This regex will have fast negatives meaning valid identifiers may not pass
+// this test. However this is only used during static generation to provide hints
+// about why a page bailed out of some or all prerendering and we can use bracket notation
+// for example while `ಠ_ಠ` is a valid identifier it's ok to print `params['ಠ_ಠ']`
+// even if this would have been fine too `params.ಠ_ಠ`
+const isDefinitelyAValidIdentifier = /^[A-Za-z_$][A-Za-z0-9_$]*$/
```

Comment:
are we duplicating these in each file to optimize for deletion?

Reply:
yes but we don't have to